File: 241025/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
n,m = map(int, input().split(" "))

visited = [
    [0 for _ in range(m)]
    for _ in range(n)
]
grid = [list(map(int, input().split())) for _ in range(n)]

def in_range(x,y):
    return 0 <= x < n and 0 <= y < m

def can_go(x,y):
    if not in_range(x,y):
        return False
    if visited[x][y] or grid[x][y] == 0:
        return False
    return True

def dfs(x,y):
    dxs, dys = [1,0], [0,1]

    visited[x][y] = 1

    for dx, dy in zip(dxs, dys):
        new_x, new_y = x+dx, y+dy

        if can_go(new_x, new_y):
            if dfs(new_x, new_y):
                return 1

    return 0

print(dfs(0,0))

File: 241102/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
#define MAX_NUM 100
#define DIR_NUM 2

using namespace std;

int n, m;
int grid[MAX_NUM][MAX_NUM];
bool visited[MAX_NUM][MAX_NUM];

bool InRange(int x, int y) {
    return 0 <= x && x < n && 0 <= y && y < m;
}

bool Cango(int x, int y) {
    if (!InRange(x, y)) {
        return false;
    }

    if (visited[x][y] || grid[x][y] == 0) {
        return false;
    }
    return true;
}

void DFS(int x, int y) {
    int dx[DIR_NUM] = {0, 1};
    int dy[DIR_NUM] = {1, 0};

    for (int i = 0; i < DIR_NUM; i++) {
        int new_x = x + dx[i];
        int new_y = y + dy[i];

        if (Cango(new_x, new_y)) {
            visited[new_x][new_y] = true;
            DFS(new_x, new_y);
        }
    }
}

int main() {
    cin >> n >> m;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> grid[i][j];
        }
    }

    visited[0][0] = true;
    DFS(0, 0);

    cout << visited[n-1][m-1];
    return 0;
}